src/psij/executors/batch/batch_scheduler_executor.py

```diff
@@ -0,0 +1,539 @@
+
+    def _read_aux_files(self, job: Job, status: JobStatus) -> None:
+        try:
+            # read exit code and output files
+            exit_code_str = self._read_aux_file(job, '.ec')
+            if exit_code_str:
+                status.exit_code = int(exit_code_str)
+                if status.exit_code != 0:
+                    status.state = JobState.FAILED
+            if status.state == JobState.FAILED:
+                if status.message is None:
+                    # only read output from submit script if another error message is not
+                    # already present
+                    status.message = self._read_aux_file(job, '.out')
+            else:
+                self._delete_aux_file(job, '.out')
+
+        except Exception as ex:
+            logger.warning('Job %s: failed to read auxiliary files: %s' % (job.id, ex))
+
+    def _read_aux_file(self, job: Job, suffix: str) -> Optional[str]:
+        assert job.native_id
+        path = self.work_directory / (job.native_id + suffix)
+        if path.exists():
+            try:
+                with open(path) as f:
+                    return f.read()
+            finally:
+                assert isinstance(self.config, BatchSchedulerExecutorConfig)
+                if not self.config.keep_files:
+                    path.unlink()
+        else:
+            return None
+
+    def _delete_aux_file(self, job: Job, suffix: str) -> None:
+        assert job.native_id
+        path = self.work_directory / (job.native_id + suffix)
+        if path.exists():
+            path.unlink()
+
+    def list(self) -> List[str]:
+        """Returns a list of jobs known to the underlying implementation.
+
+        See :func:`~JobExecutor.list`.
+        The returned list is a list of `native_id` strings representing jobs known to the
+        underlying batch scheduler implementation, whether submitted through this executor or not.
+        Implementations are encouraged to restrict the results to jobs accessible by the current
+        user.
+        """
+        pass
+
+
+class _QueuePollThread(Thread):
+    def __init__(self, name: str, config: BatchSchedulerExecutorConfig,
+                 executor: BatchSchedulerExecutor):
+        super().__init__()
+        self.name = name
+        self.daemon = True
+        self.config = config
+        self.executor = executor
+        # native_id -> job
+        self._jobs = {}  # type: Dict[str, Job]
+        # counts consecutive errors while invoking qstat or equivalent
+        self._poll_error_count = 0
+        self._jobs_lock = RLock()
+
+    def run(self) -> None:
+        logger.debug('Executor %s: queue poll thread started', self.executor)
+        while True:
+            self._poll()
+            time.sleep(self.config.queue_polling_interval)
+
+    def _poll(self) -> None:
+        with self._jobs_lock:
+            if len(self._jobs) == 0:
```

Comment:
True (pardon the pun).

Reply:
I'm inclined to keep as is since, at least for me, it's clear what `len(x) == 0` means, but for `bool(x)` I have to think: what type is x, what are all the cases when `bool(x)` is true/false, do I really want to test all of them, or just this specific thing. For example, I was surprised that `bool(JobState.COMPLETED)` was False without defining `JobState.__bool__` (i.e., I would have expected `bool(enum)` to be `is not None`, by default. Also that `__ne__` does not default to `not __eq__`, but to `False`.